### PR TITLE
Adding Open Router inference Example

### PR DESCRIPTION
**Pull request by PhiBrandon**
Simple example utlizes Openrouter API to access Open source models for inference. Utilizing Json mode within Instructor, we can get Json structured output.

OpenRouter - https://openrouter.ai/

## Summary by CodeRabbit

- **New Features**
  - Introduced a new user details extraction feature using an open source model.
  - Implemented a user-friendly interface for inputting and receiving user details.

- **Enhancements**
  - Improved user interaction by creating specific messages for data retrieval and error handling.

- **Documentation**
  - Updated documentation to reflect new features and usage instructions.

### CODE REVIEW

File: examples/open_source_examples/openrouter.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    age: int
+    occupation: Optional[str] = Field(None, description="Occupation of the user")
+
+# Utilizing teknium/openhermes-2.5-mistral-7b, we need to ensure the system prompt is present
+model = "teknium/openhermes-2.5-mistral-7b"
+system_message = {"role": "system", "content": f"You are an expert at outputting json. Match your response to this json_schema:{UserDetail.model_json_schema()['properties']}"}
+user_message = {"role": "user","content": "Extract the user details from the following text: Brandon is 33 years old. He works as a solution architect."}
+
+# Let's call the model
+user = client.chat.completions.create(response_format={"type": "json_object"}, model=model, messages=[system_message,user_message])
```
Comment (jxnl) [inline examples/open_source_examples/openrouter.py:30]: if you patched with jsonmode, and set the `response_model=UserDetail` it should return the object right?

**Comment by jxnl** [comment]
Take a look at #225 maybe we shuold add some tests or change the conftest to detect open router too

File: examples/open_source_examples/openrouter.py (hunk 1 above)
Comment (jxnl) [inline examples/open_source_examples/openrouter.py:30]: just resolve the messages into that array to make it cleaner.
please check if `response_model=UserDetail` works :D

**Review by jxnl — CHANGES_REQUESTED** [review]

File: examples/open_source_examples/openrouter.py [hunk 2]
```diff
@@ -1,38 +1,97 @@
...
-    occupation: Optional[str] = Field(None, description="Occupation of the user")
+    """Details in the text about the user"""
+    name: str = Field(description="Name extracted from the text")
+    age: int = Field(description="Age extracted from the text")
+    occupation: Optional[str] = Field(
+        default=None, description="Occupation extracted from the text")
 
-# Utilizing teknium/openhermes-2.5-mistral-7b, we need to ensure the system prompt is present
+
+def create_openai_client(api_key: str, base_url: str):
```
Comment (jxnl) [inline examples/open_source_examples/openrouter.py:30]: no need for function, this is only called once

File: examples/open_source_examples/openrouter.py [hunk 3]
```diff
@@ -1,38 +1,97 @@
...
+        "content": f"You are an expert at outputting json. You always output valid pydantic model_json_schemas."
+    }
+    try:
+        response = client.chat.completions.create(
+            response_model=Maybe(UserDetail),
+            response_format={"type": "json_object"},
+            model=model,
+            messages=[system_message, user_message]
+        )
+    except Exception as e:
```
Comment (jxnl) [inline examples/open_source_examples/openrouter.py:54]: don't catch random Exception, assume it works.

File: examples/open_source_examples/openrouter.py [hunk 4]
```diff
@@ -1,38 +1,97 @@
...
+    """
+    client = openai.OpenAI(api_key=api_key, base_url=base_url)
+    return patch(client, mode=Mode.JSON)
+
+
+def get_user_details(client, model: str, user_message: dict) -> Maybe(UserDetail):
+    """
+    Extract user details using Open soure model.
+    """
+    maybe_user_detail = Maybe(UserDetail)
```
Comment (jxnl) [inline examples/open_source_examples/openrouter.py:42]: Classes hould be SnakeCase

**Review by jxnl — CHANGES_REQUESTED** [review]
make sure you run `ruff format .` after youre done,
theres no need for so much indirection for code that is ever used once, you dont need a function

**Comment by jxnl** [comment]
no reason it shouldnt just look like this 

https://github.com/jxnl/instructor/blob/main/examples/patching/patching.py

but with `client = instructor.patch(OpenAI(...), mode=Mode.JSON)`

File: examples/open_source_examples/openrouter.py (hunk 4 above)
Comment (PhiBrandon) [reply examples/open_source_examples/openrouter.py:42]: When you say SnakeCase do you mean UserDetail should be User_Detail OR MaybeUserDetail?
I thought this was snake_case, no?

File: examples/open_source_examples/openrouter.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    user_message = create_user_message(message)
+    user = get_user_details(model, user_message)
+    # Assertions for validation
+    if user.error:
+        retry_message = create_user_message(message, user.message)
+        user = get_user_details(model, retry_message)
+        if user.result:
+            print(user.result)
+        else:
+            raise ValueError(user.message)
```
[the remark this answers is not in the document]
Comment (PhiBrandon) [reply examples/open_source_examples/openrouter.py:105]: Removing this, as we can specify retries within the patched client.

File: examples/open_source_examples/openrouter.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+            },
+            {
+                "role": "user",
+                "content": f"Extract the user details from the following text: {content}",
+            },
+        ]
+    return {
+        "role": "user",
+        "content": f"Extract the user details from the following text: {content}. Match your response to this json_schema:{MaybeUser.model_json_schema()}",
+    }
```
[the remark this answers is not in the document]
Comment (PhiBrandon) [reply examples/open_source_examples/openrouter.py:77]: Removed the error parameter, to just create the messages. This should return the system prompt with the user message as list[dict].

**Pull request merged**